### ISSUE TITLE
L1Loss / torch.abs gradient discrepancy

### ISSUE THREAD

**Issue by griegler** [pytorch/pytorch#7172]
Given the simple example below

```python
import torch
device = 'cpu'
a = torch.tensor([-1,0,1], dtype=torch.float64, device=device, requires_grad=True)
b = torch.tensor([0,0,0], dtype=torch.float64, device=device)
loss = torch.nn.functional.l1_loss(a, b, size_average=False)
loss.backward()
print(a.grad)

```
The output/gradient is `tensor([-1.,  1.,  1.], dtype=torch.float64)`. However, if the same loss is implemented with `torch.abs(a - b).sum()`, then the output/gradient is `tensor([-1.,  0.,  1.], dtype=torch.float64)`.

While any value in [-1,1] is correct in the sense of the subgradient, the latter one is in my opinion much more practical. Especially, if I want to mask values in a regression task (`ma * a; ma * b`), I do want zero gradients for those masked entries.

**Comment by elimohl** [comment]
@apaszke people usually use losses to minimize them and it's nice to have a chance to get optimal values. But with the gradient 1 at 0 for `l1_loss` we cannot reach them ever.
If you care about backward compatibility, you can add an option that changes this behavior or warning message, but I cannot think of a reason why anyone could want `1.` gradient at that point (and even more so expect that it will not be changed in the future).

**Comment by apaszke** [comment]
Honestly if your method breaks just because of this, then I could be wary of it. As mentioned before, this function is simply non-differentiable at `0`, so you might want to try a smoother alternative.

**Comment by YannDubs** [comment]
@apaszke  I agree that "you might want a smoother alternative" (btw it would be nice if we could choose the delta parameter in `SmoothL1Loss`. But the fact that it is in torch.nn documented  as "Loss" makes probably every user believe that the subgradient is 0 (even more so when we know that the torch.abs at 0 is 0) . I think the subgradients should either be modified or at least documented (or warning when instantiating?). 

If anyone sees this thread and want a quick around:

```
def l1_loss(pred, target, reduction="mean"):
    """Computes the F1 loss with subgradient 0."""
    diff = pred - target
    abs_diff = torch.abs(diff)
    if reduction == "sum":
        out = abs_diff.sum()
    elif reduction == "mean":
        batch_size = abs_diff.size(0)
        out = abs_diff.view(batch_size, -1).sum(dim=1).mean(dim=0)
    elif reduction == "none":
        out = abs_diff
    return out
```

Or using a smooth alternative:

```
def huber_loss(pred, target, delta=1e-3):
    """Computes the Huber loss."""
    diff = pred - target
    abs_diff = torch.abs(diff)
    loss = torch.where(abs_diff < delta,
                       0.5 * diff**2,
                       delta * (abs_diff - 0.5 * delta))
    return loss
```

**Comment by elimohl** [comment]
> Honestly if your method breaks just because of this, then I could be wary of it. As mentioned before, this function is simply non-differentiable at 0, so you might want to try a smoother alternative.

So, do you suggest avoiding using L1Loss for any method that have a chance to converge to 0?

**Comment by apaszke** [comment]
@YannDubs well the choice of zero is entirely subjective though. If we change it others will later come and say that they want to have some kind of a tie-breaking behavior from L1 norm, and they want the subgradient to be 1 or -1. If the zero gradient at this single point is very important for your application you can always implement a custom `autograd.Function` that explicitly zeros the elements you want.

**Comment by YannDubs** [comment]
@apaszke subjective from a mathematical point of view but not really from an API one (it is called loss not absolute value of differences). A loss is by definition something to minimize, and taking any other subgradients than 0 would increase the loss.

I agree that you should not make changes when there are multiple mathematically sound values just because someone ask you to, but I have the feeling this is different because of the function contains the name "loss" (but maybe everyone thinks their case is different ;))

I'm fine now that I know it and use a wrapper, but users might be mislead due to the naming. Maybe I should just add a note in the docs rather than complaining ;)
